Some Possible bug Fixes

Bob-IT opened the pull request:
Some possible bug fixes.
The 'blank split' one was strange as it was there in 2.6.x also. From the comment in the code, that part of code is used when another row is deleted but I could only get it to run in this situation. To be safe I added an 'if blank split' statement to not destroy the split when it is the blank split.
The Invoice search bug was caused by an update to the other search options to pass the parent window for error messages and this one was missed as it was not in the search-gnome directory. I was going to move it but it would also require other file changes that may be best dealt with on master.

Bob-IT commented:
Chris, suggestions always welcome, have changes assoc to association

gjanssens commented on libgnucash/engine/engine-interface.scm:
```diff
@@ -136,7 +136,7 @@
    "gnc:transaction-structure"
    '(transaction-guid currency date-entered date-posted
-                      num description notes split-scms)))
+                      num description notes association split-scms)))
```
I'm not sure I agree on the reasoning in your commit message. I do understand you logic and follow it to some extent.
However to my mind cut == copy + delete and duplicate == copy + paste. By only making cut transfer the associated file you're breaking this universal semantic. Which may cause more confusion.

Your reasoning also assumes the user wants to associate unique files to transactions. But that may not be the case. Like I can imagine the user wants to attach a pdf of a bank statement to all transactions associated with that particular bank statement.

I think this is something too user dependent for gnucash to decide and I would propose the paste transaction (and by extension the duplicate transaction code) checks if a file association is present and if so ask the user what to do with it: add it to the new transaction as well or not.

Bob-IT replied:
After some more thinking, I believe that if you are cutting/copying and then pasting a transaction you should expect all information for that transaction so I would propose that be the case with the associated entry. With the duplicate action, it already brings up a dialogue for cheque number and date so I think we can add to that, if there is an associated entry, add a tick box with 'Keep Associated Entry' with the default being ticked.

Bob-IT replied:
I have just noticed one other scenario, auto complete, should the associated entry be copied then, I am thinking not.

gjanssens replied:
No, autocomplete shoud not copy the associated entry.

gjanssens commented:
Ok, that's good enough for me.

gjanssens commented on gnucash/gnome/dialog-trans-assoc.c:
```diff
@@ -144,7 +144,7 @@ convert_uri_relative_to_uri (AssocDialog *assoc_dialog, const gchar *uri)
     const gchar *new_uri;
 
-    if (assoc_dialog->valid_path_head && g_str_has_prefix (uri,"file:/") &&
+    if (assoc_dialog->valid_path_head && g_str_has_prefix (uri,"file:") &&
```
You are correct https://tools.ietf.org/html/rfc8089#section-3 doesn't specify relative file uri's. However our code needs a way to handle those anyway.

Unfortunately your code has chosen a different convention from mine in gnc-uri-utils.c. This is a potential source of confusion and errors. That's what I'm pointing out. I won't claim my work is right. I have written the uri code a long time ago and most likely was pretty naive while writing it.

For starters it won't parse a localhost intermediate. It will simply become part of the local path name. This is really a bug, but I don't know if it matters in practice.

Your choice for relative file uri's is `file:/my-shared-text-file.txt` while mine is `file://my-shared-text-file.txt`. Clearly my scheme will lead to ambiguity if the relative path would be `file://hostname/my-shared-text-file.txt` so retrospect this is an unfortunate choice. If I remember correctly I made that choice to universally be able to strip `file://` from the URI to get either an absolute or a relative path.

For consistency I think we should abandon at least my choice of a relative scheme for file handling. If a file URI is passed to GnuCash, GnuCash should resolve it as if it's absolute. If that results in a "file not found" this should be reported to the user as such. On the other hand, GnuCash internally keeps track of the book's location in the form of a file URI. So if the path passed to GnuCash was relative, GnuCash can still make it absolute using some conversion routine before making it a URI. That way our internal URI is always consistent.

Similarly while looking at your use case I doubt we need to support something like `file:/somefile.txt`. It would be better to also unambiguously store either a proper file uri (meaning with absolute path) or simply a relative path without uri scheme. Your code can easily detect whether there's a uri scheme in the stored string and act accordingly. No scheme: assume this is a local path name and determine absoluteness before adding a potential path head. With file scheme, consider path always absolute. You are using the GLib uri functions to parse file schemes, so it should do the right thing for you.

Perhaps this is also what you meant with 
> (that probably should change which I was contemplating along with some other changes later)

The unfortunate part is that we currently have code in the wild that uses these contrived schemes. I haven't studies your code in much more detail, so I wonder in which circumstances will it store a relative path using the `file:/path/to/file.txt` scheme and when `path/to/file.txt` ?

Bob-IT replied:
At the moment, the stored value for absolute file paths is 'file:///usr/share/file.txt' or for a relative path either 'file:/my-file.txt' or 'file:my-file.txt' depending on what the path head was set to.

If the 'transaction location association' was changed to only accept uri's, then any thing else can be treated as a relative path so the following should work when the association is retrieved...
Test for 'file:' but not 'file://', remove the 'file:' part and save this stripped value.
Test for a scheme with g_uri_parse_scheme,

- if a scheme but not 'file' then treat as location uri
- If a file scheme then it is absolute
- If it is not a scheme, add it to what ever the path head is for an absolute path.

Over time all entries with the old format would get changed and if they used the 'Transaction Association' dialogue they would all get changed as part of the retrieval process.

jralls replied:
This is a bit skew to the discussion at hand, but I'd prefer that you use/extend the URI handling in libgnucash/engine/gnc-uri-utils.h to using GLib ones. I think those are the ones that @gjanssens is referring to. Doing so will promote better uniformity across GnuCash and make it easier to migrate away from GLib. For the code that @gjanssens originally commented on that would be ```gnc_uri_is_file_uri```; instead of ```g_uri_parse``` use ```gnc_uri_get_components```.

Bob-IT commented:
I have made some changes to the commits specified, the copy/cut/paste the associated entry I think is correct and the fix for the storing/retrieving associated files does what I think it should but may need more work.
I have added a check for when a location is specified to see if it has a valid scheme to warn the user.
Added a couple of functions to gnc_uri_ to get the scheme and build an absolute path for the file associations. These are used as appropriate in the storing/retrieving of the association.
Also added a convert function to fix the original error for relative paths so they are stored without a scheme and '/' in native form.

gjanssens commented on libgnucash/engine/gnc-uri-utils.c:
```diff
@@ -342,3 +342,81 @@ gchar *gnc_uri_add_extension ( const gchar *uri, const gchar *extension )
+gboolean gnc_uri_absolute_associate_file_path (const gchar *uri,
+                                               const gchar *path_head,
+                                               gchar **ret_uri)
+{
```
Related question: can the base path also be a location ? Your code so far only accepts a file uri as base path. However perhaps all files are on an ftp server instead. That's currently not supported and perhaps for now we should keep it that way. If we do want to support it the relative path handling may require two functions: one for local file system paths (via gnc-filepath-utils) and one for other uri based base paths.

Bob-IT replied:
Do you agree with the following, just trying to make sure I understand the starting point.
The path head will be stored in a form like...
Windows

- file://localhost/c:/somedir
- file:///c:/somedir
- file:///c:/ 

Linux

- file://locallhost/usr/somedid
- file:///usr/somedir
- file:///

A Location Association must have a valid scheme/protocol
A File Association could be relative or absolute
The stored associate value will be in the form depending on OS and whether file or location based...

- file:///c:/User/Admin/some-file.txt
- some-file.txt
- docs/list/some-file.txt
- http://www.gnucash.org
- ftp://myname@host.dom/home/myname/some-file.txt
- file:///usr/somedir/some-file.txt

This would make invalid scheme/protocols a relative path and can be made absolute by adding to path head and with a '/' depending on whether the path head ended with a '/'

Bob-IT commented on libgnucash/engine/gnc-uri-utils.c:
```diff
@@ -342,3 +342,81 @@ gchar *gnc_uri_add_extension ( const gchar *uri, const gchar *extension )
+}
+
+gchar *
+gnc_uri_parse_scheme (const gchar *uri)
```
I added this so that I could get a working solution for what I had in mind as I needed to distinguish between a 'file' protocol and a non existent one, `gnc_uri_get_protocol` would return 'file'.

Bob-IT commented:
Was there more work to do on Bug767772 ?

gjanssens commented:
Op vrijdag 16 november 2018 12:32:57 CET schreef Bob-IT:
> Was there more work to do on Bug767772 ?

Probably not. I skipped it because it was related to file associations and 
didn't look into the details. I will asap.

gjanssens commented:
Op vrijdag 16 november 2018 12:31:38 CET schreef Bob-IT:
> Bob-IT commented on this pull request.
> 
> > @@ -342,3 +342,81 @@ gchar *gnc_uri_add_extension ( const gchar *uri,
> > const gchar *extension )
>      /* Ok, all tests passed, let's add the extension */
>      return g_strconcat( uri, extension, NULL );
>  }
> +
> +gboolean gnc_uri_absolute_associate_file_path (const gchar *uri,
> +                                               const gchar *path_head,
> +                                               gchar **ret_uri)
> +{
> 
> Do you agree with the following, just trying to make sure I understand the
> starting point. The path head will be stored in a form like...
> Windows
> 
> - file://localhost/c:/somedir
> - file:///c:/somedir
> - file:///c:/
> 
> Linux
> 
> - file://locallhost/usr/somedid
> - file:///usr/somedir
> - file:///
> 
I assume it can also be empty. In which case it would make sense to use the 
location of the currently open data file as a starting point. Of course that 
doesn't work when file associations are relative and the book is in mysql of 
postgresql.
There are several possible fixes for that:
* We could come up with yet another default directory in that situation (like 
GND_DATA_HOME/data
* Refuse relative file associations (that is, warn the user if s/he entered 
one)
* Ask the user to specify a base directory the first time s/he enters a 
relative file association.

I prefer the last one. As a sidenote I now think it was unfortunate to store 
the starting point in GSettings instead of in the book. I would consider 
associations grouped per book much more sensible than grouped per user. But 
that's a change we could only make for 4.x unfortunately.

And note while file://localhost/usr/somedir is a valid uri I don't know if we 
care enough to support it. On the other hand as gnc-uri-utils already knows 
how to extract hostnames there's not added burden to support it either.

> A Location Association must have a valid scheme/protocol
> A File Association could be relative or absolute
> The stored associate value will be in the form depending on OS and whether
> file or location based...
> 
> - file:///c:/User/Admin/some-file.txt
> - some-file.txt
> - docs/list/some-file.txt
> - http://www.gnucash.org
> - ftp://myname@host.dom/home/myname/some-file.txt
> - file:///usr/somedir/some-file.txt
One more possibility:
/home/user/somefile.txt
c:\users\user\somefile.txt

That is, I think we should also support absolute filenames without a scheme. 
That's probably the most common use case. Note that file uri's should be an 
implementation detail. The less a user sees of it in the interface the better.

> 
> This would make invalid scheme/protocols a relative path and can be made
> absolute by adding to path head and with a '/' depending on whether the
> path head ended with a '/'

So no, in the absence of a scheme, the path is assumed to be on the local file 
system. It should be checked whether that path is then relative or absolute. 
If absolute, use it as is, if relative make it absolute by prepending the 
starting point (which can be either user-defined via the setting or assumed to 
be the book's current directory). If it can't be made absolute because the 
book is mysql/postgresql and there's no starting point defined, fall back to 
the suggestions above.

gjanssens commented:
Op vrijdag 16 november 2018 11:49:48 CET schreef Bob-IT:
> Bob-IT commented on this pull request.
> 
> > +                *ret_uri = g_strconcat (path_head, "/", uri, NULL);
> 
> +            else
> +                *ret_uri = g_strconcat (path_head, uri, NULL);
> +        }
> +        else
> +            *ret_uri = g_strconcat ("file:///", uri, NULL);
> +
> +        g_free (ph_scheme);
> +        return TRUE;
> +    }
> +    g_free (uri_scheme);
> +    return FALSE;
> +}
> +
> +gchar *
> +gnc_uri_parse_scheme (const gchar *uri)
> 
> I added this so that I could get a working solution for what I had in mind
> as I needed to distinguish between a 'file' protocol and a non existent
> one, `gnc_uri_get_protocol` would return 'file'.

Right. That's part of the gnc-uri-utils that needs fixing. My implementation 
was aimed purely at distinguishing between file/xml/sqlite/mysql/postgresql to 
do some backend specific stuff. There was no other use case back then. Now we 
additionally need to distinguish between a file uri and a filesystem path.

So some of the assumptions need to be revised. And the first one is probably 
what gnc_uri_get_protocol should return in the absence of a scheme and by 
extension gnc_uri_get_components.. Both should probably return NULL instead of 
defaulting to "file".

However that also means we should revisit each use of these functions and 
update the code to work with the new assumption.

I believe in many cases we can get by with a new function

gboolean gnc_uri_is_uri (const gchar* uri)

which will return whether the passed in uri has a real scheme or not. The 
other uri testing functions (like gnc_uri_is_file_uri/
gnc_uri_is_file_protocol) make no sense if that function returns FALSE. And 
they should equally return FALSE if called with a local filesystem path or a 
NULL protocol.

I have used gnc_uri_is_file_uri in the past to determine whether or not we can 
have passwords with the given uri scheme. Here also I now think this is the 
wrong part of the code to store this knowledge. This is something I believe 
the backends should be able to answer. I don't know if they currently can. If 
they can't it's something to consider in the future. For now this can stay as 
is. But given gnc_uri_is_file_uri should return FALSE if the uri passed in is 
not a uri at all, we will have to extend most calls to gnc_uri_is_file_uri 
with an additional call to gnc_uri_is_uri as we now have 3 possible cases for 
the backend related file/uri handling:
- local filename, use unchanged
- file type uri, won't have passwords, when displayed to user, only display 
uri path component
- mysql, postgresql uri: can have passwords. When displayed to user, strip 
password

Note our backends also support an xml and sqlite uri scheme. These are not 
official uri's but for our purposes they behave exactly as the file uri scheme 
from the gnc-uri point of view. The different does matter for the backends. 
For the file association starting point, you can accept them and automatically 
convert them to file:/// in my opinion. You can do so as soon as the user save 
the setting as far as I'm concerned, so that your code can always assume 
gn_uri_is_file_uri really means a file:/// uri.

gjanssens commented:
Sorry for the double posts. I have fixed a few typos and improved a few sentences. The core message is unchanged.

Bob-IT commented:
This looks like it is getting more complicated!!!

> I assume it can also be empty. In which case it would make sense to use the location of the currently open data file as a starting point.

Yes, it is empty by default, the way it is now, the user can not enter a relative path and I do not want them to. So they use the GtkFileChooserDialog to select a file to associate and this returns an absolute file uri. I then look at this uri to see if it starts with the same uri as path head, if it does, that part is removed from the uri with the remainder being saved or if it does not the whole uri is saved.

So if there is no path head, absolute file uri's are saved, I do not see why we need to change this aspect. 

> One more possibility:
 /home/user/somefile.txt
 c:\users\user\somefile.txt
 That is, I think we should also support absolute filenames without a scheme. That's probably the most common use case. Note that file uri's should be an implementation detail. The less a user sees of it in the interface the better.

How are they going to enter that ?, file selection is by the GtkFileChooserDialog. They do see the scheme of the uri in the 'Transaction Associations Dialogue' to match the location uri's format.

I will have to think about the other post, may be after a sleep.

gjanssens commented:
@Bob-IT I have started checking the uses of gnc_uri_... functions. It looks like most of data file related code internally assumes complete uris. Our GncHistory code is one big exception. That will store filesystem paths instead. I suspect I chose this because the history entries can also be looked at via the platform's GSettings backends (dconf, Windows registry, Macos defaults) and I wanted the entries visible to users to be in the system local format as much as possible. So for history interactions we'll need to complement the gnc_uri_is_file_uri with a gnc_uri_is_uri to distinguish between strings starting with file://, xml://, sqlite:// or not scheme at all.
I also still have to double-check sessions store a uri as book_id, or whether it can also be a local filesystem path. If the latter, there are two additional spots to fix this:
* gnc-state.c:100
* gnc-main-window.c:1521
Both start from the return value of qof_session_get_url, for which the comments specify it will be the url that  was opened, with local filesystem files in the form "file:/some/file.sgml". I fear this comment is slightly outdated, so I'll have to dig into the backends to get the current status. That will take me a little more time.

So overall if we extend gnc-uri-utils to return a more strict answer on gnc_uri_is_file_uri and add a supplementary gnc_uri_is_uri, the changes outside of this code are pretty limited.

gjanssens commented:
Hmm, hit post too soon...

I think for gui interaction qof_session_get_url will always return a proper answer as whatever is set as backend at some point had to pass via gnc-file (gnc_post_file_open, gnc_do_save_as). Both of these ensure a normalized uri is passed to the backend code.
I don't know however how this is handled when starting from python or guile scripts.

gjanssens commented:
@Bob-IT Are you working on this already ? If you want I can take over the gnc-uri-utils refactoring so you can focus on the transaction associations afterwards.

Bob-IT commented:
@gjanssens  I did start but other things have got in the way so please do, I was also thinking about renaming the gnc_uri_is_file... to something like gnc_uri_is_local... as it kept confusing me that it related to file/xml/sqlite3 files and then adding a gnc_uri_is_file_protocol that actually was just file protocol.

gjanssens commented:
This got sidetracked on my part. Sorry about that.

I have now cherry-picked Bug767772. So that fix will go into gnucash 3.4. I have also reworked the gnc-uri-utils interface. There is now gnc_uri_is_uri to test if the input is a true uri (with scheme, path and potentially a host name). gnc_uri_is_file_uri is a now more specialized version of that same check. It will in addition check whether the uri refers to a file system object (it won't check for the existence though). A simple local file system path (without scheme) is no longer considered a valid uri nor a file uri.
There is an additional function gnc_uri_targets_local_fs that only checks if the input refers to a local file system object. This will return TRUE for both file uris and plain local file system paths.
Note this is still not a full uri specification implementation. It's limited to the current uses of uris within gnucash and it cuts a lot of corners that don't really matter for us.
I believe with that you should be able to continue your work on the transaction associations.

jralls commented:
Is what's left of this superseded by #460?

Bob-IT commented:
Yes this can be closed when 460 is done.

jralls commented:
The remainder of this PR was committed under #460.
